### DISCUSSION TITLE
How to convert existing xcodeproj to bazel with xcodemigrate?

### DISCUSSION THREAD

**Discussion by palaniraja** [dahchon/BazelGen#37] [Q&A]
I had trouble in using the app. I am trying to setup Bazel pipeline (local) for this opensource app [IceCubesApp](https://github.com/Dimillian/IceCubesApp) by @Dimillian

Note: `➜  icecubeapp git:(main) ✗`  is my oymyzsh prompt 

### System information

```
➜  icecubeapp git:(main) ✗ sw_vers
ProductName:	macOS
ProductVersion:	12.6.2
BuildVersion:	21G320
```

```
➜  icecubeapp git:(main) ✗ xcodebuild -version
Xcode 14.2
Build version 14C18
```

Bazel was installed via homebrew

```
➜  icecubeapp git:(main) ✗ bazel --version
bazel 6.0.0-homebrew
```

```
➜  icecubeapp git:(main) ✗ which bazel
/opt/homebrew/bin/bazel
```

xcode-migrate was installed via mint

```
➜  icecubeapp git:(main) ✗ which xcode-migrate
/Users/palaniraja/.mint/bin/xcode-migrate
```

For whatever reason I wasnt able to run via `mint run` even though my  my $PATH is `export PATH="$HOME/.mint/bin:$PATH"%`

```
➜  icecubeapp git:(main) ✗ mint run xcode-migrate
🌱  "xcode-migrate" package not found
```

So i had to run just `xcode-migrate`. And readme usage is failing for me

```
➜  icecubeapp git:(main) ✗ xcode-migrate generate -p IceCubesApp.xcodeproj
Error: The value 'generate' is invalid for '<action>'
Help:  <action>  Action to execute. Available action: bazel
Usage: xcode-migrate <action> --project <project> [--config <config>]
  See 'xcode-migrate --help' for more information.
```

same for `--project`

```
➜  icecubeapp git:(main) ✗ xcode-migrate generate --project IceCubesApp.xcodeproj
Error: The value 'generate' is invalid for '<action>'
Help:  <action>  Action to execute. Available action: bazel
Usage: xcode-migrate <action> --project <project> [--config <config>]
  See 'xcode-migrate --help' for more information.
```

As you can see I have only `bazel` option available for _action_

```
➜  icecubeapp git:(main) ✗ xcode-migrate bazel -p IceCubesApp.xcodeproj
[XcodeMigrate] info: No configuration file found, using default configuration
[XcodeMigrate] info: Project path: IceCubesApp.xcodeproj
[XcodeMigrate] info: Bazel project generated at: IceCubesApp.xcodeproj
```

and no options to proceed further

```
➜  icecubeapp git:(main) ✗ bazel query //IceCubesApp:*
zsh: no matches found: //IceCubesApp:*

➜  icecubeapp git:(main) ✗ bazel query //
ERROR: the empty string is not a valid target
Loading: 0 packages loaded
```

**Comment by MikeChongCan** [comment]
Thanks for your post!

First of all, `xcode-migrate` is still in its early stage. After my local testing, I found generation for `IceCubesApp` is not fully supported. Because:

1. We haven't implement the support for Swift package. There are two kinds of Swift packages used in this project: Internal and External. I will link related issues in this post soon.
2. There's a bug about handling source file path, which needs to be fixed.
3. For this app, its Info.plist misses app version. We need to add the ability to automatically add a default version for this scenario.

Once all these issues fixed, you can generate the build definitions for this app.

Before that, I recommend you to create a simple project using [Tuist](https://tuist.io). At this stage, `xcode-migrate` will work.

If you really want to build the app and see errors from bazel, here is the command: 

```
bazel build //IceCubesApp/App:IceCubesApp
```

I highly recommend to add debug flags to see the errors:

```
bazel build //IceCubesApp/App:IceCubesApp --sanbox_debug --verbose_failures
```

I will let you know when `xcode-migrate` is ready for this project. It may take a number of weeks or months (I work for this project as a hobby, without any income from it).

**Reply by palaniraja** [reply]
I completely understand it's your hobby project.

Appreciate the post with additional info. 

I am new to Bazel and i still don't understand how it works. Saw your project was interested in converting my project to understand more.

IceCubeApp was a choice because of open source.

I'll try the options you explained and update my learning here if anyone lost as me in the future 

Thanks again for sharing project!

**Reply by MikeChongCan** [reply]
> I am new to Bazel and i still don't understand how it works. Saw your project was interested in converting my project to understand more.

If you need a straightforward example, there are several that I can mention:

- https://github.com/realm/SwiftLint
- https://github.com/TelegramMessenger/Telegram-iOS

My example:

- https://github.com/imWildCat/bazel-ios-minimal-example
